### PR TITLE
Improve error message when Proxy object is iterated

### PR DESCRIPTION
**Pull request by ansley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58302 Improve error message when Proxy object is iterated**

Differential Revision: [D28444030](https://our.internmc.facebook.com/intern/diff/D28444030)

### CODE REVIEW

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444030).

File: torch/fx/proxy.py [hunk 1]
```diff
@@ -134,8 +134,13 @@ def iter(self, obj: 'Proxy') -> Iterator:
...
         """
-        raise TraceError('Proxy object cannot be iterated. '
-                         'This can be attempted when used in a for loop or as a *args or **kwargs function argument.')
+        raise TraceError('Proxy object cannot be iterated. This can be '
+                         'attempted when the Proxy is used in a loop or'
+                         ' as a *args or **kwargs function argument. '
+                         'Consider factoring out the untraceable logic '
+                         'using torch.fx.wrap. See the torch.fx docs on'
+                         ' pytorch.org for a more detailed explanation '
+                         'of what types of control flow can be traced.')
```
Comment (jamesr66a) [inline torch/fx/proxy.py:143]: I think there's another alternative: if the control flow is static (e.g. the loop trip count is based on some hyperparameter), the code can be refactored as something like:

```
for i in range(self.some_hyperparameter):
  indexed_item = proxied_value[i]
```

We might want to work something like that in as well

**Review by jamesr66a — APPROVED** [review]

**Comment by ansley** [comment]
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28444030).

File: torch/fx/proxy.py (hunk 1 above)
Comment (ansley) [reply torch/fx/proxy.py]: Good idea! I moved some of that information up into the Proxy docstring to give it more visibility. Please check over the new things I added before I land the fixes

**Review by jamesr66a — APPROVED** [review]

**Pull request closed without merging**